tket/src/Circuit/CircUtils.cpp

```diff
@@ -266,139 +267,69 @@ std::pair<Circuit, Complex> decompose_2cx_DV(const Eigen::Matrix4cd &U) {
   return {circ, std::conj(z0)};
 }
 
-Circuit phase_gadget(unsigned n_qubits, const Expr &t, CXConfigType cx_config) {
-  // Handle n_qubits==0 as a special case, or the calculations below
-  // go badly wrong.
-  Circuit new_circ(n_qubits);
-  if (n_qubits == 0) {
-    new_circ.add_phase(-t / 2);
-    return new_circ;
+Expr pauli_angle_convert_or_throw(Complex pauliCoeff, const Expr &angle) {
+  if (pauliCoeff == -1.) {
```

Comment:
Whenever QubitPauliTensor is used as part of Tableau classes, diagonalisation methods, gadget methods, etc., the coefficient is always in the discrete value set {1, i, -1, -i}. All operations throughout preserve this value set (really, the imaginaries only occur as intermediate values for easy multiplication and we should semantically only have +-1 at most points). Repeatedly checking floating-points within some tolerance seems a bit cumbersome for something we could statically assert with better typing. I'm now thinking a better solution might be to bake these discrete values into the class, e.g. QubitPauliStabiliser a sparse variant of the PauliStabiliser class whose coeff is an enum rather than a bool. Would you agree or would this be overkill?

Reply:
Yeah I did wonder about that. I agree, if they're guaranteed to be one of those discrete values let's bake that into the type (perhaps storing an int representing the power of i?).

Reply:
Implemented with the new specialisations of PauliTensor from #1081 which have been merged into this